Merging Properties and POM based configuration

stromnet opened the pull request:
(originally discussed in #938. Branch based on commit in #941 with run/buildEnv)

This adds supports to use both Properties based configuration and POM based configuration to configure the same image. This is done by adding the `source` configuration to the `<external>` section, which can have one of the following values:
* `properties` - default, same as existing properties based configuration (ie. no merge)
* `pom, properties` - merge pom configs and properties, with values in POM having priority
* `properties, pom` - merge pom configs and properties, with values in properties having priority
* (`pom` - pretty useless, same as not having <external> at all)

Can also be enabled by pure properties (only makes sense for single image), by setting the property `docker.externalPropertyConfiguration.source` to a valid `source` value. This is useful if we want to add properties from site-wide config for example settings.xml, for example adding `runEnv.http_proxy` without touching every POM using the plugin.

Properties are loaded from Maven Project properties merged with System properties (Maven -D flags).

Merging on value level takes place for only a few specific properties by default, for most values it does not make sense to merge (i.e. `runCmd`, `from` or any other single values).
* For Lists, merging happens by appending all values to one list.
* For Maps, merging happens by adding all keys to one map. Source with higher priority overwrites keys with lower priority.

Which values to be merged/replaced are controlled by a new configuration on the `ConfigKey` enum. This can be overridden by setting the `_combine` property suffix.

h2. Examples
```
<properties>
   <docker.network.alias.1>mycontainer</docker.network.alias.1>
</properties>
....
<image>  
  <external>
    <type>properties</type>
    <source>properties,pom</source> <!-- This one is new -->
  </external>
  <name>myimage</name>
  <build> 
       ...
  </build>
   <run>
      ....
   </run>
</image>
```

Set some properties through -D flags (requires that external is activated in POM)
```
mvn docker:run -Ddocker.runEnv.http_proxy=http://my.runtime.specific.proxy.com:1234
```

Set some properties through -D flags, also activating external from -D flag:
```
mvn docker:run -Ddocker.externalPropertyConfiguration.source=properties,pom \
  -Ddocker.runEnv.http_proxy=http://my.runtime.specific.proxy.com:1234
```

Override merge policy for runEnv map (only use the one defined in properties, ignoring any in POM):
```
mvn docker:run -Ddocker.externalPropertyConfiguration.source=properties,pom \
  -Ddocker.runEnv.http_proxy._combine=replace \
  -Ddocker.runEnv.http_proxy=http://my.runtime.specific.proxy.com:1234
```

stromnet commented:
Ping @rhuss

rhuss commented:
Awesome, thanks ! I'm on the road this week (and was on PTO last week), so I probably can review the PR only at the beginning of next week.

'hope this is ok ...

stromnet commented:
Great, OK! 
If you could take a quick glance at the above description of functionality (not review of actual implementation) it would be great, if acceptable I can start preparing our projects for using this.

rhuss commented:
Looks great, thanks again ! I like the way how `<source>` is implemented.

Some minor nits about the naming: I would probably made it more explicit that it is about naming, and `pom` is probably a too generic name (as properties are also part of the pom and can e.g. definied in a `<properties>` section).

What do you thinks about a `<mergePolicy>` tag with the possible values:

* `none`
* `properties` (or `properties,config`, maybe we can support both. Or any other idea for a catchy name here ?)
* `config` (or `config,properties`)
* `skip`

?

Not 100% sure about the _combine suffix (and whether I understand it on a first read ;-). Do you have a use case handy where this required (and can't be solved otherwise) ?

stromnet commented:
Hm yes I was going back and forth with `pom` vs `config` vs something else..  Agree that PM is a bit too ambigious. Other suggestions?

Are you suggesting having mergePolicy instead of `source` (to indicate which order to merge stuff)? I.e. to set `<mergePolicy>properties</mergePolicy>` in order to merge existing config with properties, and properties have priority?

The `source` as it is now is only there to dictate which source to prioritize, pom/config vs properties. 
The `_combine` is different, and allows overriding of what I call CombinePolicy in the implementation, which dictates what to do with a ConfigKey when it sees multiple values.

Basically, each ConfigKey has a CombinePolicy assigned to it, for most it's  Replace. That means the highest prioritised value will be used without any sort of merge.
But for some, i.e runEnv, it has CombinePolicy Merge, which allows those values to be merged. For runCmd it does not make sense to have Merge for example, but for runEnv it does, at least we could assume so generally. But if a user actually want's to replace their full runEnv, they can set a property `docker.runEnv._combine=replace` to force the Replace policy to be used on the runEnv. Same thing with labels etc.

The defaults for the different ConfigKeys are certainly up for discussion. For example, labels, ulimit, and a few others could very well make sense to use Merge by default.

rhuss commented:
> Are you suggesting having mergePolicy instead of source (to indicate which order to merge stuff)? I.e. to set <mergePolicy>properties</mergePolicy> in order to merge existing config with properties, and properties have priority?

yes, I think has its about merging configurations maybe this should be reflected in the name of the config option, too. But if you like `<source>` better, I'm fine with that, too. Could be also 'propertiesFirst' (but that sounds a bit weird these days, too ;-)

I see the point with specifying how to combine values when merging, not 100% sure about how we expose this to the user. Let's discuss this next week, but you can assume that the PR will get in dmp asap (except maybe for minor tweaks), so that you could start migrating your projects already.

stromnet commented:
I'm thinking that first step is "sourcing values" i.e. from config and/or properties, in some order, and second step is "combining values" to get the final configuration, where "combine" is in some cases merge and in some cases replace.. Although replace might be bad word too, since "replace" indicates that we are replacing something (when all we do is really taking the value with highest priority).

One thing which is not covered is *clearing* values. For example if we have some property in config, which we want to unset with properties. In some cases a blank string might work but..

Anyway, great, looking forward to further discuss next week and merge :)

stromnet commented:
@rhuss Hi, had any time to look at this yet? :)

rhuss commented:
@stromnet yeah, sorry. Back from PTO and conference, and already kneedeep in work again ;-(

I thought a bit about the naming and what do you think about this suggestion:

```xml
 <external>
    <type>properties</type>
    <mode>only|override|fallback|skip</mode>
  </external>
```

The advantage would be

* `properties` is mentioned only once (could otherwise lead to confusion between `<type>` and `<source>`
* It's short (e.g. not commas) and hopefully intuitive. The `<mode>` describes the role of the handler its own POV with respect to the other configuration.
* Avoids confusion between `combining` and `merging`. 

I don't mind much that we don't have yet support for clearing values. We should tackle this when it becomes a requirement (i.e. a valid use case exists).

wdyt ? If we agree on the naming I would jump into a reviewer later today or tomorrow.

stromnet commented:
@rhuss welcome back :)

Good point on `properties` vs `<type>` confusion!

So, `<mode>only</mode>` would be same as todays functionality (and default).
`<mode>override</mode>` would use both, and let properties override any image configuration set in POM. 
`<mode>fallback</mode>` would use both, and only use properties if not set in image configuration set in POM. 

What would `skip` do? But sounds good to me!

It doesn't address the override/fallback scenarios where we want to do a merge, for example a env map where we want to merge both properties from POM and add some from properties. Or are you happy with the "combine policy" proposal?

rhuss commented:
> What would skip do? But sounds good to me!

`skip` just would skip using properties (your original `pom` source). Might be useful to disable it temporarily of for certain profiles. People like to have `skip` hooks I found out ;-)

> Or are you happy with the "combine policy" proposal?

Do you mean the `..._combine` specification ? I'm fine with that (and dont have another idea).

stromnet commented:
> skip just would skip using properties (your original pom source). Might be useful to disable it temporarily of for certain profiles. People like to have skip hooks I found out ;-)

Ah, of course. Yes!

> 
> Or are you happy with the "combine policy" proposal?
> 
> Do you mean the ..._combine specification ? I'm fine with that (and dont have another idea).

Ok, great!


So, existing code should be changed a bit then, but mostly around the `ValueSource` class and how it is referenced & used I think. I'll have time to fix this on Friday, so if you want you can still review the other parts of the PR tomorrow, the ValueSource specific parts are quite small. Then I can give a updated PR on Friday?

rhuss commented:
Sounds like a plan ;-) thanks !

stromnet commented on src/main/asciidoc/inc/external/_property_configuration.adoc:
```diff
@@ -94,10 +94,16 @@ when a `docker.from` or a `docker.fromExt` is set.
-| Sets an environment variable. E.g. `<docker.env.JAVA_OPTS>-Xmx512m</docker.env.JAVA_OPTS>` sets the environment variable `JAVA_OPTS`. Multiple such entries can be provided. This environment is used both for building images and running containers. The value cannot be empty but can contain Maven property names which are resolved before the Dockerfile is created.
+| Sets an environment variable used in build and run. E.g. `<docker.env.JAVA_OPTS>-Xmx512m</docker.env.JAVA_OPTS>` sets the environment variable `JAVA_OPTS`. Multiple such entries can be provided. This environment is used both for building images and running containers. The value cannot be empty but can contain Maven property names which are resolved before the Dockerfile is created.
+
+| *docker.buildEnv.VARIABLE*
```
Had the same thought. Problem is, IIRC, that the existing code looks for properties beginning with`docker.env` prefix. Consequently, if I had a property `docker.envRun.JAVA_OPTS=-Xmx512m` it was adding  `Run.JAVA_OPTS=-Xmx512m` in both run and build env.. Or at least something along those lines, don't recall exactly.
Could perhaps be fixed by altering property matching?

stromnet commented on src/main/asciidoc/inc/external/_property_configuration.adoc:
```diff
@@ -94,10 +94,16 @@ when a `docker.from` or a `docker.fromExt` is set.
 
 | *docker.envPropertyFile*
-| specifies the path to a property file whose properties are used as environment variables. The environment variables takes precedence over any other environment variables specified.
+| specifies the path to a property file whose properties are used as environment variables in run. The environment variables takes precedence over any other environment variables specified.
```
Simply because it did not exist in build and I did not add implementation for that.. ;)  `getEnvPropertyFile()` is only defined in RunImageConfiguration.

stromnet commented on src/main/asciidoc/inc/external/_property_configuration.adoc:
```diff
@@ -94,10 +94,16 @@ when a `docker.from` or a `docker.fromExt` is set.
 | Property part for the exposed container properties like internal IP addresses as described in <<start-overview, docker:start>>.
```
Check, after final solution is fully decided

stromnet replied:
Added

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java:
```diff
@@ -41,165 +40,167 @@ public String getType() {
-                .filter(withPrefix(prefix, FILTER, properties))
-                .user(withPrefix(prefix, USER, properties))
-                .healthCheck(extractHealthCheck(prefix, properties))
+                .cmd(valueProvider.getString(CMD, config == null ? null : config.getCmd()))
```
Hm, thought about alternatives to that. That is one way, but then we have to repeat the full Builder structure, right?

I guess Java8 and Optional with lambdas isn't an option?

rhuss replied:
Why would we need to rebuild the builder structure ? Don't really see this (as the members of the builder is already initialized with null).

d-m-p started as a Java 7 project at that time and hence doesn't use optionals. But I think in our case it would help quite a bit, so if you have a good suggestion, we I wouldn't mind to introduce Java 8 features.

stromnet replied:
Just extracting config != null check would look something like this:
```
if(config != null) {
    return new RunImageConfiguration.Builder()
        .capAdd(valueProvider.getList(CAP_ADD, config.getCapAdd()))
        .capDrop(valueProvider.getList(CAP_DROP, config.getCapDrop()))
        .securityOpts(valueProvider.getList(SECURITY_OPTS, config.getSecurityOpts()))
        .cmd(extractArguments(valueProvider, CMD, config.getCmd()))
        .....
        .build();
}else{
    return new RunImageConfiguration.Builder()
        .capAdd(valueProvider.getList(CAP_ADD, null))
        .capDrop(valueProvider.getList(CAP_DROP, null))
        .securityOpts(valueProvider.getList(SECURITY_OPTS, null))
        .cmd(extractArguments(valueProvider, CMD, null))
        .....
        .build();
}
```

A bit cleaner to see,, but easier to miss when adding new properties.

With java8 and lambdas, it could be something like this (Optionals or local fn):
```
static <T,R> R orNull(T config, java.util.function.Function<T,R> fn) {
    if(config != null) {
       return fn.apply(config);
    }
    return null;
}
....
return new RunImageConfiguration.Builder()
    .capAdd(valueProvider.getList(CAP_ADD, orNull(config, RunImageConfiguration::getCapAdd)))
    .capDrop(valueProvider.getList(CAP_DROP, Optional.of(config).map(RunImageConfiguration::getCapDrop).orElse(null)))
    ...
    .build();
```

Not sure if the RunImageConfiguration:: can somehow be interfered from T.. But a lot cleaner.

Anyway, adding Java8 as a requirement would hopefully not be an issue for users of the plugin, last public update of Java 7 was 2015 it seems?... At least we're not stuck in Java 6.. :) 

**BUT**,  I've seen far too many cases when Java version updates irritates, and I really really would not want to wait for any such disputes to clear up before a new release is made.. So unless you insist on updating, and still having a pretty quick release, I prefer ugly over Java 8.

stromnet replied:
* writing comment to not hide as outdated*

rhuss replied:
Ok, lets keep it that way and switch over to Optional soon after. I don't mind to set Java 8 as minimal requirement, and we are not yet 1.0.0 so should be easy todo.

And you are right Java 7 should be dead now, especially for build tools.

stromnet commented on src/main/java/io/fabric8/maven/docker/config/ImageConfiguration.java:
```diff
@@ -60,6 +69,9 @@ public RunImageConfiguration getRunConfiguration() {
         return (run == null) ? RunImageConfiguration.DEFAULT : run;
     }
 
+    public RunImageConfiguration getRunConfigurationRaw() {
```
So, in `PropertyConfigHandler` (calling code) we would change

```
RunImageConfiguration run = extractRunConfiguration(fromConfig.getRunConfigurationRaw(), valueProvider);
```

to something like this?

```
RunImageConfiguration existingRun = fromConfig.getRunConfigurationRaw()
RunImageConfiguration run = extractRunConfiguration(existingRun.isDefault() ? null : existingRun, valueProvider);
```

rhuss replied:
something like that, yes, but would do the check probably im `extractRunConfiguration()`

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java:
```diff
@@ -41,165 +40,167 @@ public String getType() {
-                .imagePullPolicy(withPrefix(prefix, IMAGE_PULL_POLICY_RUN, properties))
-                .ulimits(extractUlimits(prefix, properties))
-                .tmpfs(listWithPrefix(prefix, TMPFS, properties))
+                .capAdd(valueProvider.getList(CAP_ADD, config == null ? null : config.getCapAdd()))
```
I think problem with using RunConfiguration.DEFAULT in general was that some options would "seem" to be configured (their default values), which would affect overriding functionality. For example if config is set to override properties, and a particular property is set in property, we wouldn't want the default value (which wasn't actually set in config) to override the property. Hence we must have some way to see the difference on "set by default" and "set by user".

rhuss replied:
yeah, I see your point. Its all about having default values backed into the configuration (see one of my later comments).

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java:
```diff
@@ -41,165 +40,167 @@ public String getType() {
+                .domainname(valueProvider.getString(DOMAINNAME, config == null ? null : config.getDomainname()))
+                .entrypoint(valueProvider.getString(ENTRYPOINT, config == null ? null : config.getEntrypoint()))
+                .env(CollectionUtils.mergeMaps(
+                        valueProvider.getMap(ENV_RUN, config == null ? null : config.getEnv()),
```
If i understand correct: change `getEnv` to `getEnvRun`? If so, agree!

rhuss replied:
Ah, sorry. Misunderstood the code. I'm fine with it as it is now (as its RunImageConfiguration.getEnv() and hence clear what it is about)

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/ValueProvider.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    public String getString(ConfigKey key, String pomValue) {
+        return stringValueExtractor.getFromPreferedSource(prefix, key, pomValue);
```
Doh, IntellIJ even highlighted that.. :)

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/ValueProvider.java:
```diff
@@ -0,0 +1,277 @@
+            List<T> values = new ArrayList<>();
+
+            // Find all non-null values, with same order as sources
+            for (ValueSource source : sources) {
```
My thought too!

stromnet commented on src/main/java/io/fabric8/maven/docker/config/AssemblyConfiguration.java:
```diff
@@ -71,6 +76,10 @@
     @Parameter
     private String tarLongFileMode;
 
+    public Boolean getExportTargetDir() {
```
It seems to have logic to handle deprecated configuration modes, my thought was to use as bare values as possible (same as the added `xxxRaw` methods)

stromnet commented on src/main/java/io/fabric8/maven/docker/service/ContainerTracker.java:
```diff
@@ -199,8 +199,8 @@ private void clearAllMaps() {
             WaitConfiguration waitConfig = runConfig != null ? runConfig.getWaitConfiguration() : null;
-            this.shutdownGracePeriod = waitConfig != null ? waitConfig.getShutdown() : 0;
-            this.killGracePeriod = waitConfig != null ? waitConfig.getKill() : 0;
+            this.shutdownGracePeriod = waitConfig != null ? waitConfig.shutdown() : 0;
```
Makes sense, since that one is only used there.

stromnet commented:
Great review! Lots of good points, have commented a few of them. Will get on to fix the no-discussion ones in the meantime.

stromnet commented:
Ok, so as always, lots of stuff came in the way today and I did not have the time to do as much as I wanted... But just pushed some minor fixes at least. May be short on time next week, so not sure how much I'm going to be able to finish..

rhuss commented:
@stromnet thanks for the feedback ! No worries, I'm currently also very busy, but happy to pick up as soon as you're done.

thanks !

stromnet commented:
Got some time over, pushed most (all?) minor changes. And one inline-comment on null vs Java 8 above.

stromnet commented on src/main/java/io/fabric8/maven/docker/config/handler/property/ConfigKey.java:
```diff
@@ -119,22 +123,34 @@
     WATCH_INTERVAL("watch.interval"),
     WATCH_MODE("watch.mode"),
     WATCH_POSTGOAL("watch.postGoal"),
+    WATCH_POSTEXEC("watch.postExec"),
```
All of these seems to lack documentation but leaving that to someone else.. :)

rhuss commented:
cool, thanks ! I'm just on the road for Javaland, but hopefully I find some train-time to make the final review. Let's get it merged asap ;-)

stromnet commented:
@rhuss There, I think I've addressed all issues now I think, except alternative solution for repeated `config == null` checks, see that comment (unfortunately hidden as "outdated").

rhuss commented:
just a heads up: I'm going to merge the PR this afternoon when I'm back

stromnet commented:
@rhuss Great :) Any objections to me squashing everything to one commit and re-pushing it before you do that?

stromnet commented:
Rebased on master and re-pushed (same branch) with squashed commits.
The original branch if you want to compare it, is on https://github.com/stromnet/docker-maven-plugin/tree/property-merging-nosquash

rhuss commented:
did a quick re-review of your changes and all looks good. Thanks a lot !

So, let's merge it now. 

However, for the next steps I really would like to streamline the default handling (i.e. how to provide default values, maybe getting rid of the "raw" methods and null checks). 
We should use the full Java 8 machinery for this for sure. 

Not sure how to tackle this, but would be super happy if you could give it a spin.

Thanks again a ton, it's really a super valuable feature. It does not look big, but make this plugin much more flexible.

stromnet commented:
Perfect, thanks! Eagerly waiting for the release then :)

One oops: noticed my squashed commit msg mentions the wrong property.
